### DISCUSSION TITLE
Runtime validation of output on client and server

### DISCUSSION THREAD

**Discussion by carnesen** [trpc/trpc#1517] [RFC / Ideas]
We are evaluating tRPC and want to confirm one aspect of its feature set that we found surprising. Can you confirm that tRPC does not currently support automatic runtime validation of route output on client and server using Zod schema? What I mean is that I want to define a route like 

```
export const appRouter = trpc
  .router()
  .query("user", {
    input: z.string(),
    output: z.object({
      userId: z.string(),
      name: z.string(),
    }),
    async resolve({ input: userId }) {
      return await db.findUser(userId);
    },
  });
```
and tRPC will automatically:
- At compile time, type-check that the return type of my resolver is compatible with the type of the output specification.
- At runtime on the server, parse the object returned by the resolver using the `output` Zod schema
- At runtime on the client, parse the object returned by the server using the `output` Zod schema

We'd like to do runtime validation of both the input and the output on both the client and server because the schemas can get out of sync for example if we've deployed a new server but the client is still running an older version of the schema.

**Comment by KATT** [comment]
Unsure why you'd want runtime validation on the client, you should be able to trust results coming from your own server, right?

If you want runtime validation on the output, you can check out this thread I wrote  on twitter https://twitter.com/alexdotjs/status/1488148385353895939

**Reply by KATT** [reply]
I'd be open to add an optional `output:` validator if someone from the community would like to put in the work for it.

**Reply by carnesen** [reply]
Thank you for the reply. I saw your note on Twitter and the linked Gist. That would help us do server-side runtime output validation. Since we aren't able to ensure that our client and server software are deployed in tandem, the client can't trust the data coming from the server. The server might have been upgraded in a way that's not backwards compatible with the version of the software running on the client. We want such incompatibilities to manifest as hard parse errors. Basically we take the stance that all data entering the process should be runtime-type-parsed, never simply type-asserted. Unless you suggest otherwise, I'll leave this open as a feature request and as you say perhaps someone from the community will pick it up.

**Reply by KATT** [reply]
Thanks for sharing the context. I'm going to give this some time to marinate before answering ☺

**Reply by mmkal** [reply]
@carnesen I'm with you on wanting an optional `output` field on resolvers, but would that actually help your client trust the server? If your server upgraded in a backwards-incompatible way, would then output validation succeed sinc ethe schema was upgraded too? Or are you talking about loading the server's zod types in the client at build time?

**Reply by carnesen** [reply]
If the server upgrades in a backwards-incompatible way, I want output schema validation to fail immediately on the client until it upgrades too. For a concrete example suppose the output Zod schema is
```
const outputSchema = z.object({
  userID: z.string(),
});
```
Now some diligent engineer comes along and says, "Hey that property name isn't properly camel-cased. The D in userID should be lower-cased". The server engineer changes the schema to:

```
const outputSchema = z.object({
  userId: z.string(),
});
```
updates the resolver, and deploys. Problem solved. OR IS IT?!? The client is still running with the OLD version of the output schema. Since the client isn't validating the output schema at runtime, they're now receiving an output object of shape `{ userId: string }` and asserting its type as `{ userID: string }`. Somewhere downstream in their code they do something that expects `userID` is a `string` such as `output.userID.split('-')` and they get a difficult-to-diagnose runtime error `TypeError: Cannot read property 'split' of undefined`. Far better to raise an error immediately where that data enters the process with an easy to understand message like what Zod would raise:

```
{
  "code": "invalid_type",
  "expected": "string",
  "received": "undefined",
  "path": [ "userID" ],
  "message": "required"
}
```

**Reply by KATT** [reply]
This seems a bit like baking infra/deployment issues into the code. Either way, it's an unrecoverable error until the front-end is deployed and the end-user has reloaded the page. 

Alternative approaches, if you can orchestrate the deployments to be timely:

- Your React error boundary could compare the running commit hash to the latest commit hash and prompt the user to reload the page 
- You can have your own `<Link/>` component that periodically check for app updates and switches a real browser navigation rather than using the clientside router (which should gracefully download new bundles) 


I'm sure it would be possible to find a way to guarantee this as part of the framework, but if you're going to force the devs to write a schema for every procedure to cater for this, you might be better off using a toolset that is catered towards this sort of strictness, like GraphQL or OpenAPI. Declaring explicit outputs for every procedure sort of ruins the point of tRPC a bit.

**Reply by carnesen** [reply]
> This seems a bit like baking infra/deployment issues into the code.

If we were able to consistently deploy client and server concurrently it would definitely lessen the need for strict runtime type checking. In any case though we parse to a schema all data entering the process, no matter how confident we are that we know its type. Here is the blog post that got us going down that path https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ .

> you might be better off using a toolset that is catered towards this sort of strictness, like GraphQL

Indeed that's what we use for browser-server communication, and we love the strictness. We got excited about tRPC because it shares the organizational philosophy of GraphQL (query, mutation, subscription) but is lighter weight and pure TypeScript. It's always seemed silly that with GraphQL we define a non-TypeScript type layer between two TypeScript applications. (We use Angular on the front end and Node.js server-side).

> Declaring explicit outputs for every procedure sort of ruins the point of tRPC a bit.

We actually prefer to be explicit about our output/return types for code readability. We've turned on the ["explicit module boundary" ESLint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md). Within our codebase it's not really an extra burden to declare a Zod schema in place of a regular type either because with our "don't trust, parse" philosophy we already have Zod schemas around for all our entities.

Thank you for your contributions to open-source software. Very nice package you've got here.

**Reply by KATT** [reply]
This thread will be interesting to you: https://discord.com/channels/867764511159091230/894984069291581511/980940295006998589

We discussed different approaches that might be relevant to you.

**Comment by jlalmes** [comment]
An `output` validation property was added to procedures in release `v9.21.0`, please see: https://github.com/trpc/trpc/releases/tag/v9.21.0
